File: Strategy Summit/2D Dynamic Programming/Q3. Word Break/Q3_WordBreak.py
class Solution(object):
    def wordBreak(self, s, wordDict):
        """
        :type s: str
        :type wordDict: List[str]
        :rtype: bool
        """
        word_set = set(wordDict)  # convert list to set for fast lookup
        n = len(s)
        dp = [False] * (n + 1)    # dp[i] = True if s[0:i] can be segmented
        dp[0] = True  # empty string can always be segmented

        # Check all prefixes
        for i in range(1, n + 1):
            for j in range(i):
                # If s[0:j] can be segmented and s[j:i] is in dictionary
                if dp[j] and s[j:i] in word_set:
                    dp[i] = True
                    break  # no need to check further for this i

        return dp[n]  # return if the entire string can be segmented

File: Strategy Summit/2D Dynamic Programming/Q3. Word Break/Q3_WordBreak.java
import java.util.*;

class Solution {
    public boolean wordBreak(String s, List<String> wordDict) {
        Set<String> wordSet = new HashSet<>(wordDict); // convert list to set for fast lookup
        boolean[] dp = new boolean[s.length() + 1];    // dp[i] = true if s[0..i-1] can be segmented
        dp[0] = true;  // empty string can always be segmented

        // Check all prefixes of s
        for (int i = 1; i <= s.length(); i++) {
            for (int j = 0; j < i; j++) {
                // If s[0..j-1] can be segmented and s[j..i-1] is in dictionary
                if (dp[j] && wordSet.contains(s.substring(j, i))) {
                    dp[i] = true;
                    break; // no need to check further for this i
                }
            }
        }

        return dp[s.length()]; // return if entire string can be segmented
    }
}
